vignettes/crunch-internals.Rmd

```diff
@@ -37,4 +37,65 @@ datasets(proj) <- ds
+
+*Real cube*: The cube with all of the underlying data
+*User cube*: The cube the user thinks they’re interacting with. This is the same as the real cube without MR selection dimensions
+*Printed cube* The cube which prints to the console. The same as the the user cube but with missing categories suppressed. 
+
+The user cube differs from the real cube in that it doesn’t include MR selection dimensions. We assume that when the user is including an MR in a cube, they really only care about the parts of the cube where the MR is selected. So the 5d real cube listed above would become a 3d user cube with 300 entries (`5 * 6 * 10`) because we are selecting the slices of the cube where each MR selection dimension is equal to “Selected” and are able to drop the selections dimensions
+
+The printed cube differs from the user cube in that it doesn’t include missing categories. We assume that the user mostly wants to look at categories which are not-missing, and that they don’t care that much about the parts of the cube which represent missing categories. If the categorical variable above had two missing categories: “No Data” and “Not Answered” it would reduce the number of entries to 240 (`5 * 6 * 8`) . This behavior can be changed by setting the `cube@useNA` slot to `“always”`.  
+
+
+### Cube subsetting
+
+This structure gets a little tricky when we subset the cube. For instance the cube we’ve been working with has 5 dimensions, but to the user appears to only have 3, so should they subset it as a 5 dimensional array like `cube[1, 1,  2:3, 1, 5]` or a 3 dimensional array `cube[1, 2:3, 5]`. Similarly, if a dimension is missing, should that be included in the subset. For example if a cube has three entries A, B, C, along one dimension, and B is hidden, should `cube[ , 1:2]` return A, A & B, or A &C? Similarly should `dim(cube)` show the dimensions of the real cube or the printed cube? 
+
+We resolve this problem by allowing the user to select whether they want the cube to represent the user cube or the printed cube, and make all of the behavior of those cubes consistent. The `cube@useNA` controls whether the user wanted to interact with the printed cube or the user cube and they can `showMissing()` and `hideMissing()` functions to control which cube they are interacting with 
+
+```
+> dim(cube)
+[1] 5 6 8
+> dim(showMissing(cube))
+[1] 5 6 10
+```
+
+This makes subsetting a bit easier because we can enforce the dimensions of the printed or literal cube. For instance
+
+```
+cube[5, 6, 10] 
+```
+ Errors because the user supplied the wrong number of dimensions while
+
+```
+showMissing(cube)[cube[5, 3, 6, 3, 10] 
+```
```

Comment:
I think this line would still error, shouldn't it be `showMissing(cube)[5, 6, 10]`?